File: docs/src/test/java/org/springframework/security/docs/servlet/oauth2/resourceserver/jwtgrantedauthoritiesspelexpression/ExpressionJwtGrantedAuthoritiesConverterTests.java
package org.springframework.security.docs.servlet.oauth2.resourceserver.jwtgrantedauthoritiesspelexpression;

import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;

import org.junit.jupiter.api.Test;

import org.springframework.expression.Expression;
import org.springframework.expression.spel.standard.SpelExpressionParser;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.jwt.TestJwts;
import org.springframework.security.oauth2.server.resource.authentication.ExpressionJwtGrantedAuthoritiesConverter;

import static org.assertj.core.api.Assertions.assertThat;

class ExpressionJwtGrantedAuthoritiesConverterTests {

	@Test
	public void convertWhenTokenHasCustomClaimNameExpressionThenCustomClaimNameAttributeIsTranslatedToAuthorities() {
		// @formatter:off
		Jwt jwt = TestJwts.jwt()
				.claim("nested", Collections.singletonMap("scopes", Arrays.asList("read", "write")))
				.build();
		// @formatter:on
		// tag::spel-expression[]
		SpelExpressionParser parser = new SpelExpressionParser();
		Expression expression = parser.parseExpression("[nested][scopes]");
		ExpressionJwtGrantedAuthoritiesConverter converter = new ExpressionJwtGrantedAuthoritiesConverter(expression);
		Collection<GrantedAuthority> authorities = converter.convert(jwt);
		// end::spel-expression[]
		assertThat(authorities).extracting(GrantedAuthority::getAuthority)
				.containsExactly("SCOPE_read", "SCOPE_write");
	}
}

File: docs/src/test/kotlin/org/springframework/security/kt/docs/servlet/oauth2/resourceserver/jwtgrantedauthoritiesspelexpression/ExpressionJwtGrantedAuthoritiesConverterTests.kt
package org.springframework.security.kt.docs.servlet.oauth2.resourceserver.jwtgrantedauthoritiesspelexpression

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.tuple
import org.junit.jupiter.api.Test
import org.springframework.expression.spel.standard.SpelExpressionParser
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.oauth2.jwt.TestJwts
import org.springframework.security.oauth2.server.resource.authentication.ExpressionJwtGrantedAuthoritiesConverter

class ExpressionJwtGrantedAuthoritiesConverterTests {
	@Test
	fun convertWhenTokenHasCustomClaimNameExpressionThenCustomClaimNameAttributeIsTranslatedToAuthorities() {
		// @formatter:off
		val jwt = TestJwts.jwt()
				.claim("nested", mapOf("scopes" to listOf("read", "write")))
				.build()
		// @formatter:on
        // tag::spel-expression[]
        val parser = SpelExpressionParser()
		val expression = parser.parseExpression("[nested][scopes]")
		val converter = ExpressionJwtGrantedAuthoritiesConverter(expression)
		val authorities = converter.convert(jwt)
        // end::spel-expression[]
		assertThat(authorities).extracting(GrantedAuthority::getAuthority)
            .containsExactly(tuple("SCOPE_read"), tuple("SCOPE_write"))
	}
}
